File: typescript/wordy/wordy.ts
type Operations = {
  [key: string]: Function
}

const operations: Operations = {
  plus: (x: number, y: number): number => x + y,
  minus: (x: number, y: number): number => x - y,
  multiplied: (x: number, y: number): number => x * y,
  divided: (x: number, y: number): number => x / y
}

const validate_question = (question: string): void => {
  if(question.match(/cubed/)) throw Error('Unknown operation')
  if(!question.match(/What is/)) throw Error('Unknown operation')
}

const validate_stack = (stack: Array<string|number>): void => {
  if(!stack.length) throw Error('Syntax error')
  if(stack.length % 2 === 0) throw Error('Syntax error')
  stack.forEach((elem, index) => {
    if(typeof elem !== 'string' && index % 2 === 1) throw Error('Syntax error')
  })
}

export const answer = (question: string): number => {
  validate_question(question)
  let re = /(-*\d+|plus|minus|multiplied|divided)/g
  let stack = question.match(re)?.map(test => parseInt(test) || test) || []
  validate_stack(stack)

  let result = stack?.shift()
  while(stack?.length) {
    let op = stack?.shift() || 'n/a'
    let num = stack?.shift()
    result = operations[op](result, num)
  }

  return Number(result)
}

File: ruby/wordy/wordy.rb
class WordProblem
  # OPERATIONS_HASH maps the operation to the corresponding method
  # @return [Hash<String, Symbol>] the operation to method mapping
  OPERATIONS_HASH = {
    "plus" => :+,
    "minus" => :-,
    "multiplied" => :*,
    "divided" => :/
  }.freeze

  # initialize method
  # @param question [String]
  def initialize(question)
    @valid = valid_question?(question)
    @digits = question.scan(/-?\d+/).map(&:to_i)
    @operations = question.scan(/plus|minus|multiplied|divided/)
  end

  # answer method returns the memoized result of the calculate method
  # @return [Integer]
  def answer
    raise ArgumentError unless @valid

    @answer ||= calculate
  end

  private

  # calculate method performs the operations on the digits
  # @return [Integer]
  def calculate
    result = @digits.shift
    while @operations.any?
      opp = OPERATIONS_HASH[@operations.shift]
      result = result.send(opp, @digits.shift)
    end
    result
  end

  # valid_question? method checks if the question is valid_question
  # ensures that the question contains at least one digit
  # and no ivalid operations
  # @param question [String] the question to validate
  # @return [Boolean]
  def valid_question?(question)
    question.match?(/-?\d+/) && !question.match?(/cubed|raised/)
  end
end
